[JSC] Fix SExt8To64 and SExt16To64 in B3ReduceStrength

hyjorc1 opened the pull request:
#### 1854261519f6bb8943925a5075229076567af26f
<pre>
[JSC] Fix SExt8To64 and SExt16To64 in B3ReduceStrength
<a href="https://bugs.webkit.org/show_bug.cgi?id=257389">https://bugs.webkit.org/show_bug.cgi?id=257389</a>
rdar://109826986

Reviewed by Yusuke Suzuki.

We need to insert the newly added values to B3 graph otherwise they
will become orphaned values.

* JSTests/wasm/stress/b3-signed-extend-16-to-64.js: Added.
(async let):
* JSTests/wasm/stress/resources/b3-signed-extend-16-to-64.wasm: Added.
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:

Canonical link: <a href="https://commits.webkit.org/264606@main">https://commits.webkit.org/264606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4b34968b180687623cf0060012efe8d36116183

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8365 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9732 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8275 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11034 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8222 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9301 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9853 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7392 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14967 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6892 "Failed to checkout and rebase branch from PR 14401") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7725 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10872 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7645 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7989 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6499 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8244 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7299 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1883 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1948 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11508 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8466 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7730 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2034 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -1884,13 +1884,11 @@ class ReduceStrength {
-                                m_index, m_value->origin(), mask & 0x7f)));
+                    Const32Value* maskValue = m_insertionSet.insert<Const32Value>(m_index, m_value->origin(), mask & 0x7f);
+                    Value* bitAndValue = m_insertionSet.insert<Value>(m_index, BitAnd, m_value->origin(), input, maskValue);
+                    replaceWithNew<Value>(ZExt32, m_value->origin(), bitAndValue);
```
Why do we need the ZExt32? Isn’t that guaranteed from the BitAnd constant?

kmiller68 replied:
Oh duh because the left operand is I32. Never mind me.

hyjorc1 replied:
This is because B3::Value is typed. Since `m_value->child(0)->child(1)->hasInt32()`, that `BitAnd` is `Int32`. But `SExt8To64` wants a `Int64` result.

hyjorc1 committed:
[JSC] Fix SExt8To64 and SExt16To64 in B3ReduceStrength
https://bugs.webkit.org/show_bug.cgi?id=257389
rdar://109826986

Reviewed by Yusuke Suzuki.

We need to insert the newly added values to B3 graph otherwise they
will become orphaned values.

* JSTests/wasm/stress/b3-signed-extend-16-to-64.js: Added.
(async let):
* JSTests/wasm/stress/resources/b3-signed-extend-16-to-64.wasm: Added.
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:

Canonical link: https://commits.webkit.org/264606@main
